.github/workflows/post-pr-comment.yml

```
@@ -17,15 +22,36 @@ jobs:
         env:
           GITHUB_TOKEN: ${{ github.token }}
           WORKFLOW_RUN_ID: ${{ github.event.workflow_run.id }}
-      - run: |
-          PR="$(grep -o '^[0-9]\+$' pr.txt)"
+
+      - name: Check that PR SHA matches workflow SHA
+        run: |
+          PR="$(grep -o '^[0-9]\+$' pr_number.txt)"
           PR_HEAD_SHA="$(gh api "/repos/${GITHUB_REPOSITORY}/pulls/${PR}" --jq .head.sha)"
           # Check that the pull-request head SHA matches the head SHA of the workflow run
           if [ "${WORKFLOW_RUN_HEAD_SHA}" != "${PR_HEAD_SHA}" ]; then
             echo "PR head SHA ${PR_HEAD_SHA} does not match workflow_run event SHA ${WORKFLOW_RUN_HEAD_SHA}. Stopping." 1>&2
             exit 1
           fi
-          gh pr comment "${PR}" --repo "${GITHUB_REPOSITORY}" -F comment.txt
         env:
           GITHUB_TOKEN: ${{ github.token }}
           WORKFLOW_RUN_HEAD_SHA: ${{ github.event.workflow_run.head_commit.id }}
+
+      - name: Create or update comment
+        run: |
+          PR_NUMBER="$(grep -o '^[0-9]\+$' pr_number.txt)"
+          # comment_id.txt may be empty if there is no existing comment
+          if [ -s comment_id.txt ]
+          then
+            COMMENT_ID="$(grep -o '^[0-9]\+$' comment_id.txt)"
+          fi
+
+          if [ -z "$COMMENT_ID" ]
+          then
+            # Create new comment
+            jq --rawfile body comment_body.txt '{"body":$body}' -n | gh api "repos/${GITHUB_REPOSITORY}/issues/${PR_NUMBER}/comments" -X POST --input -
+          else
+            # Update existing comment
+            jq --rawfile body comment_body.txt '{"body":$body}' -n | gh api "repos/${GITHUB_REPOSITORY}/issues/comments/${COMMENT_ID}" -X PATCH --input -
+          fi
```

There is no need for a double `if`. The case where `comment_id.txt` exists but does not contain a number is an error condition that should never happen. I think the workflow wouldn't even fall through to the `-z "${COMMENT_ID"` check because `grep` fails with a non-zero exit code if the pattern isn't matched.

```suggestion
          if [ -s comment_id.txt ]
          then
            COMMENT_ID="$(grep -o '^[0-9]\+$' comment_id.txt)"
            # Update existing comment
            jq --rawfile body comment_body.txt '{"body":$body}' -n | gh api "repos/${GITHUB_REPOSITORY}/issues/comments/${COMMENT_ID}" -X PATCH --input -
          else
            # Create new comment
            jq --rawfile body comment_body.txt '{"body":$body}' -n | gh api "repos/${GITHUB_REPOSITORY}/issues/${PR_NUMBER}/comments" -X POST --input -
          fi
```

---

Unfortunately, the `jq` expression to select the ID of the most recent matching comment (`[ ... ] | max`) will return `null` if there are no matching comments, and `gh` will for some reason render that as an empty line. This means we end up with a `comment_id.txt` containing just a newline, which bash considers a non-empty file (i.e. `-s comment_id.txt == true`).

The approach I've gone with is to catch the grep failure and then use `[ $COMMENT_ID ]` as the test.